source/controllers/include/controllers/ControllerFactory.hpp

```diff
@@ -0,0 +1,81 @@
+#pragma once
+
+#include "controllers/IController.hpp"
+#include "controllers/ControllerType.hpp"
+#include "robot_model/Model.hpp"
+
+#include "state_representation/space/cartesian/CartesianState.hpp"
+#include "state_representation/space/joint/JointState.hpp"
+
+namespace controllers {
+
+/**
+ * @class ControllerFactory
+ * @brief Factory to create a shared pointer to a controller.
+ * @tparam S Underlying state type of the controller
+ */
+template<class S>
+class ControllerFactory {
+public:
+
+  /**
+   * @brief Create a controller of the desired type.
+   * @param type The type of controller
+   * @param dimensions The dimensionality of the controller (6 for Cartesian space, number of joints for joint space)
+   * @return The shared pointer to the controller
+   */
+  static std::shared_ptr<IController<S>> create_controller(CONTROLLER_TYPE type, unsigned int dimensions = 6);
+
+  /**
+   * @brief Create a controller of the desired type with initial parameters
+   * @param type The type of controller
+   * @param parameters A list of parameters to set on the controller
+   * @param dimensions The dimensionality of the controller (6 for Cartesian space, number of joints for joint space)
+   * @return The shared pointer to the controller
+   */
+  static std::shared_ptr<IController<S>> create_controller(
+      CONTROLLER_TYPE type, const std::list<std::shared_ptr<state_representation::ParameterInterface>>& parameters,
```

Comment:
I think you could reduce those constructors to two only by setting a default value to empty list for this list of parameters:

```cpp
static std::shared_ptr<IController<S>> create_controller(
      CONTROLLER_TYPE type, unsigned int dimensions = 6,
      const std::list<std::shared_ptr<state_representation::ParameterInterface>>& parameters=std::list<std::shared_ptr<state_representation::ParameterInterface>>()
);
```

This way you can also have the same in the robot model version with the parameter list always being the last element.

```cpp
static std::shared_ptr<IController<S>> create_controller(
      CONTROLLER_TYPE type, const robot_model::Model& robot_model,
      const std::list<std::shared_ptr<state_representation::ParameterInterface>>& parameters=std::list<std::shared_ptr<state_representation::ParameterInterface>>()
  );
```

It is a bit more verbose but remove two function declarations in the process.

Reply:
I see. Indeed there is not really a good way of doing that then. Keep it like this then it does not adds so much more overhead anyway. I would have preferred to have params last (hence why I suggested that) as it avoid the order change, but because dimensions has a default value you can't have both at the time.